route_rules: add "state: absent" support

ffmancera opened the pull request:
Signed-off-by: Fernando Fernandez Mancera <ffmancera@riseup.net>

cathay4t reviewed (changes requested):
You patch only allows removing specific route rule without support of wildcards.

Please refer to route entry for example implementation.

ffmancera committed:
route_rules: add "state: absent" support

Signed-off-by: Fernando Fernandez Mancera <ffmancera@riseup.net>

ffmancera commented:
@cathay4t This is ready for review now. I've added more integration tests

cathay4t reviewed (changes requested):
```yml
---
routes:
  config:
  - destination: 192.168.2.0/24
    metric: 108
    next-hop-address: 192.168.1.3
    next-hop-interface: eth1
    table-id: 200
  - destination: 2001:db8:a::/64
    metric: 108
    next-hop-address: 2001:db8:1::2
    next-hop-interface: eth1
    table-id: 200
route-rules:
  config:
    - ip-from: 192.168.3.2/32
      route-table: 200
    - ip-from: 2001:db8:b::/64
      route-table: 200
interfaces:
  - name: eth1
    type: ethernet
    state: up
    mtu: 1500
    ipv4:
      enabled: true
      dhcp: false
      address:
      - ip: 192.168.1.1
        prefix-length: 24
    ipv6:
      enabled: true
      dhcp: false
      autoconf: false
      address:
      - ip: 2001:db8:1::1
        prefix-length: 64
```

```yml
---
route-rules:
  config:
    - state: absent
      route-table: 200
interfaces:
  - name: eth1
```
Does not works.

ffmancera committed:
nm.ipv6: call clear_routing_rules() when creating the setting

In order to remove IPv6 routing rules Nmstate needs to call
clear_routing_rules() on the IPv6 setting as it is done on the IPv4
setting.

Added a testcase for this.

Signed-off-by: Fernando Fernandez Mancera <ffmancera@riseup.net>

ffmancera commented:
> ```yaml
> ---
> routes:
>   config:
>   - destination: 192.168.2.0/24
>     metric: 108
>     next-hop-address: 192.168.1.3
>     next-hop-interface: eth1
>     table-id: 200
>   - destination: 2001:db8:a::/64
>     metric: 108
>     next-hop-address: 2001:db8:1::2
>     next-hop-interface: eth1
>     table-id: 200
> route-rules:
>   config:
>     - ip-from: 192.168.3.2/32
>       route-table: 200
>     - ip-from: 2001:db8:b::/64
>       route-table: 200
> interfaces:
>   - name: eth1
>     type: ethernet
>     state: up
>     mtu: 1500
>     ipv4:
>       enabled: true
>       dhcp: false
>       address:
>       - ip: 192.168.1.1
>         prefix-length: 24
>     ipv6:
>       enabled: true
>       dhcp: false
>       autoconf: false
>       address:
>       - ip: 2001:db8:1::1
>         prefix-length: 64
> ```
> 
> ```yaml
> ---
> route-rules:
>   config:
>     - state: absent
>       route-table: 200
> interfaces:
>   - name: eth1
> ```
> 
> Does not works.

This was happening because the interface were not being marked as changed therefore, if the user set the minimum required state it was not working. Now the interface is being marked as changed. Test case added.

In addition, I've found a bug on IPv6 route rules. I've created a fix and an integration test in a separated commit because it needs to be backported.

cathay4t reviewed (changes requested):
All good.

Please add unit test cases testing the wildcard match on all supporting propertied not just
route table id.

ffmancera commented:
> All good.
> 
> Please add unit test cases testing the wildcard match on all supporting propertied not just
> route table id.

Done. I've added 3 new cases testing all the features with wildcard. I hope that is good :-)

cathay4t commented on libnmstate/route_rule.py:
```diff
@@ -84,38 +102,64 @@ def is_valid(self, config_iface_routes):
+                    new_rules.add(rule)
+            if new_rules != rule_set:
+                self._rules[route_table] = new_rules
+                for iface in ifaces.values():
+                    if iface.is_ipv4_enabled or iface.is_ipv6_enabled:
+                        iface.mark_as_changed()
```
You are marking all interfaces as changed regardless they are holding route rule or not.

Since at this point, we don't know which interface will be used for saving this route rule, I think you can do
it in the `nm` level where you know current route rules and desire route rules or do it in `RouteRuleState.gen_metadata()`

ffmancera replied:
Done, it is being marked in `RouteRuleState.gen_metadata()`. Thanks and nice catch!

cathay4t commented on libnmstate/route_rule.py:
```diff
@@ -162,6 +202,7 @@ def gen_metadata(self, route_state, ifaces):
             for rule in rules:
                 family = Interface.IPV6 if rule.is_ipv6 else Interface.IPV4
                 route_rule_metadata[iface_name][family].append(rule.to_dict())
+                ifaces[iface_name].mark_as_changed()
```
With `self._cur_rules`, you can check whether desired rule changed from current rule for this table id, and only mark as changed if so.

Meanwhile, if someone removed the all rule from certain table ID, that interface will not be marked as changed. But this is little bit complex and can wait in another PR.

ffmancera commented on libnmstate/route_rule.py:
```diff
@@ -159,6 +199,8 @@ def gen_metadata(self, route_state, ifaces):
                 Interface.IPV4: [],
                 Interface.IPV6: [],
             }
+            if rules != self._cur_rules[route_table]:
+                ifaces[iface_name].mark_as_changed()
```
That handles the case if all the rules has been removed from a table. The structure `self._rules` is {table: [rules]}, therefore when all the rules has been removed it will be `{table: []}` and the generated metadata will be {iface_name: {ipv4: [], ipv6: []}} which means that all the rules has been removed.

In addition, the code is comparing self._rules with self._cur_rules for that table, so if it changed the interface will be marked as changed.

cathay4t reviewed (changes requested):
Please use `RouteRuleState.RULES_CHANGED = True` instead of two constants.

ffmancera commented:
> Please use `RouteRuleState.RULES_CHANGED = True` instead of two constants.

That is not possible. We need to identify it per interface not for the whole RouteRuleState. I've replaced it for a "_changed" = True.
